src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs

```diff
@@ -0,0 +1,1562 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Globalization;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Microsoft.CodeAnalysis.CSharp.Emit;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    internal abstract class SourcePropertySymbolBase : PropertySymbol, IAttributeTargetSymbol
+    {
+        /// <summary>
+        /// Condensed flags storing useful information about the <see cref="SourcePropertySymbolBase"/>
+        /// so that we do not have to go back to source to compute this data.
+        /// </summary>
+        [Flags]
+        private enum Flags : byte
+        {
+            IsExpressionBodied = 1 << 0,
+            IsAutoProperty = 1 << 1,
+            IsExplicitInterfaceImplementation = 1 << 2,
+        }
+
+        // TODO (tomat): consider splitting into multiple subclasses/rare data.
+
+        private readonly SourceMemberContainerTypeSymbol _containingType;
+        private readonly string _name;
+        private readonly SyntaxReference _syntaxRef;
+        protected readonly DeclarationModifiers _modifiers;
+        private readonly ImmutableArray<CustomModifier> _refCustomModifiers;
+        private readonly SourcePropertyAccessorSymbol _getMethod;
+        private readonly SourcePropertyAccessorSymbol _setMethod;
+        private readonly TypeSymbol _explicitInterfaceType;
+        private readonly ImmutableArray<PropertySymbol> _explicitInterfaceImplementations;
+        private readonly Flags _propertyFlags;
+        private readonly RefKind _refKind;
+
+        private SymbolCompletionState _state;
+        private ImmutableArray<ParameterSymbol> _lazyParameters;
+        private TypeWithAnnotations.Boxed _lazyType;
+
+        /// <summary>
+        /// Set in constructor, might be changed while decoding <see cref="IndexerNameAttribute"/>.
+        /// </summary>
+        protected readonly string _sourceName;
+
+        private string _lazyDocComment;
+        private string _lazyExpandedDocComment;
+        private OverriddenOrHiddenMembersResult _lazyOverriddenOrHiddenMembers;
+        private SynthesizedSealedPropertyAccessor _lazySynthesizedSealedAccessor;
+        private CustomAttributesBag<CSharpAttributeData> _lazyCustomAttributesBag;
+
+        // CONSIDER: if the parameters were computed lazily, ParameterCount could be overridden to fall back on the syntax (as in SourceMemberMethodSymbol).
+
+        public Location Location { get; }
+
+        protected SourcePropertySymbolBase(
+           SourceMemberContainerTypeSymbol containingType,
+           Binder bodyBinder,
```

Comment:
>Binder bodyBinder, [](start = 11, length = 18)

We should find a way to remove hard requirement on the binder. If it is hard to remove the parameter, at least it should be possible to pass null when there is nothing to bind.